src/Integration/LocalServices/TestProjectIndicators/TestProjectIndicator.cs

```diff
@@ -0,0 +1,45 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace SonarLint.VisualStudio.Integration
+{
+    internal class TestProjectIndicator : ITestProjectIndicator
+    {
+        private readonly ITestProjectIndicator buildPropertyIndicator;
+        private readonly IEnumerable<ITestProjectIndicator> testProjectIndicators;
+
+        public TestProjectIndicator(ITestProjectIndicator buildPropertyIndicator, IEnumerable<ITestProjectIndicator> testProjectIndicators)
```

Comment:
Thinking about the `bool?` problem; at the moment, this class knows that there is something special about the first `ITestProjectIndicator` and that it needs to be treated differently (although it doesn't know or care whether it is a `BuildPropertyIndicator`, so I don't think the first parameter should be called `buildPropertyIndicator`).

A couple of different approaches to consider:
1) leave the class as it is, but change the name of the first parameter to indicate its special importance e.g. `primaryTestProjectIndicator`.

2) change the meaning of the `ITestProjectIndicator` class so that all project indicators can be treated the same i.e. the indicator classes should return `null` to mean `don't know`, and only return `true` or `false` if they are sure.
The indicators should be evaluated by this class in order, and it should stop evaluating as soon as one of them returns true or false. If none of them return a value then it should return false.

Effectively, this means that `ProjectCapabilityTestProjectIndicator`, `ProjectNameTestProjectIndicator` and `ProjectKindTestProjectIndicator` should all return either `null` or `true`. They should never return `false` (because if the specific value they are looking for isn't present then they have to return "don't know" i.e. null).

Reply:
Sounds good to me, I'll make the changes.

Reply:
I made the changes but it looks a bit off to me that all the `ITestProjectIndicators` effectively return `true `or `null`.
Thinking out loud: `buildPropertyIndicator` is the only one that can definitively say "no, this is not a test project", while the others can only say "no, this is not a test project according to my definition". So maybe we should separate the interfaces, something like:
`ITestProjectMarkerChecker { bool HasTestMarker(Project project); }`
and 
`ITestProjectIndicator { bool? IsTestProject(Project project); }`

`buildPropertyIndicator` will inherit from `ITestProjectIndicator`, while the others will inherit from `ITestProjectMarkerChecker`.
And the main composite `TestProjectIndicator` can receive a list of primary `ITestProjectIndicator`, and a list of `ITestProjectMarkerChecker`. 
Wdyt?

Reply:
It seems a bit over-complicated: one interface for something that can return _yes/no/don't know_, and a different interface that can return only _yes/don't know_.

I don't have a strong preference. It's only a small amount of code and the impact is localized so it would be easy to change later.